### PR TITLE
VAN-18655 document custom fields api

### PR DESCRIPTION
**Pull request by ericmontag**

### CODE REVIEW

**Comment by ericmontag** [comment]
@shaisachs @JacksonianDemocracy @NGPVAN/falcon 
Please review.

**Comment by ericmontag** [comment]
@shaisachs @JacksonianDemocracy @NGPVAN/falcon 
Please review.

File: docs/van-api/custom-fields/common-models.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+| **`customFieldId`** | **int** | Unique identifier for a Custom Field in this context |
+| **`customFieldParentId`** | **int** | Identifies a dropdown&rsquo;s parent (for hierarchical dropdowns only) |
+| **`customFieldName`** | **string** | The name for the Custom Field, no longer than 255 characters |
+| **`customFieldGroupId`** | **int** | Unique identifier for the Custom Field Group to which this Custom Field belongs |
+| **`customFieldGroupName`** | **string** | The name of the Custom Field Group to which this Custom Field belongs, no longer than 60 characters |
+| **`customFieldTypeId`** | **string** | A one character identifier of the Custom Field type |
+| **`isEditable`** | **boolean** | Indicates whether on not the value assigned to the Custom Field can be changed |
+| **`isRange`** | **boolean** | Indicates whether the value assigned to the Custom Field can be the start of or the end of a range of values |
+| **`maxTextboxCharacters`** | **int** | The maximum number of character allowed in the text field. Only applies to Text fields |
+| **`availableValues`** | **array** | And array of dropdown options. Only applies to Dropdown fields |
```
Comment (BaconSoap) [inline docs/van-api/custom-fields/common-models.md]: should be `An`

File: docs/van-api/custom-fields/common-models.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+|:---------|:-----|:------------|
+| **`customFieldId`** | **int** | Unique identifier for a Custom Field in this context |
+| **`customFieldParentId`** | **int** | Identifies a dropdown&rsquo;s parent (for hierarchical dropdowns only) |
+| **`customFieldName`** | **string** | The name for the Custom Field, no longer than 255 characters |
+| **`customFieldGroupId`** | **int** | Unique identifier for the Custom Field Group to which this Custom Field belongs |
+| **`customFieldGroupName`** | **string** | The name of the Custom Field Group to which this Custom Field belongs, no longer than 60 characters |
+| **`customFieldTypeId`** | **string** | A one character identifier of the Custom Field type |
+| **`isEditable`** | **boolean** | Indicates whether on not the value assigned to the Custom Field can be changed |
+| **`isRange`** | **boolean** | Indicates whether the value assigned to the Custom Field can be the start of or the end of a range of values |
+| **`maxTextboxCharacters`** | **int** | The maximum number of character allowed in the text field. Only applies to Text fields |
```
Comment (BaconSoap) [inline docs/van-api/custom-fields/common-models.md]: "characters" instead of "character"

File: docs/van-api/custom-fields/common-models.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+| Property | Type | Description |
+|:---------|:-----|:------------|
+| **`customFieldId`** | **int** | Unique identifier for a Custom Field in this context |
+| **`customFieldParentId`** | **int** | Identifies a dropdown&rsquo;s parent (for hierarchical dropdowns only) |
+| **`customFieldName`** | **string** | The name for the Custom Field, no longer than 255 characters |
+| **`customFieldGroupId`** | **int** | Unique identifier for the Custom Field Group to which this Custom Field belongs |
+| **`customFieldGroupName`** | **string** | The name of the Custom Field Group to which this Custom Field belongs, no longer than 60 characters |
+| **`customFieldTypeId`** | **string** | A one character identifier of the Custom Field type |
+| **`isEditable`** | **boolean** | Indicates whether on not the value assigned to the Custom Field can be changed |
```
Comment (BaconSoap) [inline docs/van-api/custom-fields/common-models.md]: 'or' instead of 'on'

File: docs/van-api/custom-fields/common-models.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+| **`isRange`** | **boolean** | Indicates whether the value assigned to the Custom Field can be the start of or the end of a range of values |
+| **`maxTextboxCharacters`** | **int** | The maximum number of character allowed in the text field. Only applies to Text fields |
+| **`availableValues`** | **array** | And array of dropdown options. Only applies to Dropdown fields |
+| **`availableValues.id`** | **int** | The value of the dropdown option |
+| **`availableValues.name`** | **string** | The text of the dropdown option, no longer than 100 characters |
+| **`availableValues.parentValueId`** | **int** | The id of the the dropdown option in the Custom Field&rsquo;s parent dropdown |
```
Comment (BaconSoap) [inline docs/van-api/custom-fields/common-models.md]: there's some inconsistent capitalization on Dropdown vs. dropdown here - not sure which is right though

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 5]
```diff
@@ -0,0 +1,122 @@
+extends ../../../../app/views/method
+
+
+block submeta
+  - var model		= 'Custom Fields'
```
Comment (BaconSoap) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: this has tabs instead of spaces

File: docs/van-api/custom-fields/overview.jade [hunk 6]
```diff
@@ -0,0 +1,15 @@
+extends ../../../app/views/sub-topic
+
+block subtopic-meta
+  - var subtopicTitle = "Overview"
+
+block content
+  :markdown
+   Contact Custom Fields allow you to store data that is important to you about a person
+   when standard fields and other data storage options like Activist Codes are not the right tools to store that information.
```
Comment (BaconSoap) [inline docs/van-api/custom-fields/overview.jade]: "Contact Custom Fields allow you to store data about a person when standard fields and other data storage options like Activist Codes are not the right tools to store that information" - reads a bit cleaner to me. Doesn't have the word 'important' though

File: docs/van-api/custom-fields/overview.jade [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+   Contact Custom Fields belong to Groups to facilitate organization. For example, you may have a Contact Custom Field Group
+   called &quot;Language&quot; that contains fields such as &quot;Primary Language&quot; and &quot;Speaks English?&quot; to track a person&rsquo;s language preferences.
+   Contact Custom Fields can be any of the following types: Checkbox (boolean), Currency, Date, Dropdown, Number, or Text.
+   Dropdown fields have a set list of acceptable values.
+
+   Use this endpoint to retrieve information about all available Contact Custom Fields.
```
Comment (BaconSoap) [inline docs/van-api/custom-fields/overview.jade]: I think that 'these endpoints' might be a bit more accurate than 'this endpoint' since there's a few of them

File: docs/van-api/people/common-models.jade [hunk 8]
```diff
@@ -132,7 +138,13 @@ block content
     | **`type`** | **string** | Required; a known external identifier that is available in the current context. Use `votervanid` for VoterFile VANID if making a MyCampaign call.  Use `dwid` for Catalist DWIDs in either VoterFile or MyCampaign mode. |
     | **`externalId`** | **string** | Required |
 
-  a#match-responses
+    ##### CustomFieldValue
+    | Property | Type | Description |
+    | :---------|:-----|:------------|
+    | **`customField`** | **object** | [Custom Field](#custom-fields-common-models) |
+    | **`assignedValue`** | **string** | The value to be applied to the Custom Field. The value should be consistent with the type of the Custom Field |
```
Comment (BaconSoap) [inline docs/van-api/people/common-models.jade]: I'd change 'should' to 'must'

File: docs/van-api/people/methods/md/get-person-response.md [hunk 9]
```diff
@@ -60,7 +60,17 @@ If a result is found, it will look like:
       "isPreferred": null
     }
   ],
-  "identifiers": null
+  "identifiers": null,
+  "customFieldValues": [
+    {
+      "customField": { },
+      "assignedValue": "true"
```
Comment (BaconSoap) [inline docs/van-api/people/methods/md/get-person-response.md]: do we return `"true"` the string or `true` the boolean? same question with the numbers below

**Comment by BaconSoap** [comment]
Looking pretty good! I'd still hold off for @shaisachs et al

File: docs/van-api/custom-fields/common-models.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
+
+* CustomFieldTypes:
+    * **B** = Boolean (Checkbox)
+    * **S** = Selection (Dropdown)
+    * **T** = Text
+    * **D** = Date
+    * **N** = Number (Integer)
```
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: Are floats not accepted?

File: docs/van-api/custom-fields/common-models.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    * **M** = Money (Currency)
+
+```json
+{
+    "customFieldId": 167,
```
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: Chane to a bogus value like 999, 456, etc.

File: docs/van-api/custom-fields/common-models.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    "customFieldParentId": null,
```
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: Add a sample value, also bogus, e.g. 789.

File: docs/van-api/custom-fields/common-models.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    "customFieldName": "Dropdown Example",
+    "customFieldGroupId": 64,
```
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: Another bogus value.

File: docs/van-api/custom-fields/common-models.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        "id": 4,
+        "name": "No way",
+        "parentValueId": null
+      },
+      {
+        "id": 5,
+        "name": "Yes",
+        "parentValueId": null
+      }
+    ]}
```
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: Indentation on the closing brace

File: docs/van-api/custom-fields/common-models.md (hunk 10 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: @iantowns  @JacksonianDemocracy I believe not. Only integers.

File: docs/van-api/custom-fields/common-models.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    "customFieldId": 167,
+    "customFieldParentId": null,
+    "customFieldName": "Dropdown Example",
+    "customFieldGroupId": 64,
+    "customFieldGroupName": "Example Custom Field Group",
+    "customFieldTypeId": "S",
+    "isEditable": true,
+    "isRange": false,
+    "maxTextboxCharacters": 0,
+    "availableValues": [
```
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: I'd prefer if the available values did not seem like they belong to a survey question.  If you want these can be sample answers to "Favorite ice cream flavor" or what have you.  A more realistic example, if you prefer, might be "Education level", with options like "High school diploma", "College diploma", etc.

File: docs/van-api/custom-fields/common-models.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+        "name": "Yes",
+        "parentValueId": null
+      }
+    ]}
+```
+
+| Property | Type | Description |
+|:---------|:-----|:------------|
+| **`customFieldId`** | **int** | Unique identifier for a Custom Field in this context |
+| **`customFieldParentId`** | **int** | Identifies a dropdown&rsquo;s parent (for hierarchical dropdowns only) |
```
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: Since the type is called "selection", the text should be "Identifies the parent of a custom field (for selection fields with hierarchical values only.)"

File: docs/van-api/custom-fields/common-models.md [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+| **`customFieldName`** | **string** | The name for the Custom Field, no longer than 255 characters |
+| **`customFieldGroupId`** | **int** | Unique identifier for the Custom Field Group to which this Custom Field belongs |
+| **`customFieldGroupName`** | **string** | The name of the Custom Field Group to which this Custom Field belongs, no longer than 60 characters |
+| **`customFieldTypeId`** | **string** | A one character identifier of the Custom Field type |
+| **`isEditable`** | **boolean** | Indicates whether on not the value assigned to the Custom Field can be changed |
+| **`isRange`** | **boolean** | Indicates whether the value assigned to the Custom Field can be the start of or the end of a range of values |
```
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: Does the custom field data contract identify what the start and end of the range is?  If not - we should probably tack on some work to report out those values.

File: docs/van-api/custom-fields/common-models.md (hunk 4 above)
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: Good catch.  Let's go with capitalization all the way through, and stick to "Selection" since that's what the code for the type field stands for.

File: docs/van-api/custom-fields/common-models.md [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+| **`maxTextboxCharacters`** | **int** | The maximum number of character allowed in the text field. Only applies to Text fields |
+| **`availableValues`** | **array** | And array of dropdown options. Only applies to Dropdown fields |
+| **`availableValues.id`** | **int** | The value of the dropdown option |
```
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: The id, name, and parentValueId properties should all be formatted in a separate table, similar to the way the "Emails" sub-object is formatted in the People common model, see https://github.com/NGPVAN/developers.ngpvan.com/blob/master/docs/van-api/people/common-models.jade#L98

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+block response
+  :markdown
+    This endpoint responds with all Custom Fields that are available in the current context.
+    This is an example response.
+    ```json
+    [
+      {
+        "customFieldId": 164,
+        "customFieldParentId": null,
+        "customFieldName": "Checkbox Example",
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: I think it makes sense to provide names that illustrate a somewhat realistic reason why one might use such a field.  (Though in this case I grant you, there's a fine line between checkbox custom fields and activist codes.)  Sample field here might be: "Is a retiree"

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+        "customFieldTypeId": "B",
+        "isEditable": true,
+        "isRange": false,
+        "maxTextboxCharacters": 0,
+        "availableValues": []
+      },
+      {
+        "customFieldId": 165,
+        "customFieldParentId": null,
+        "customFieldName": "Currency Example",
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Same comment about realistic field names.  For ex.: "Amount of stipend"

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+    This endpoint responds with all Custom Fields that are available in the current context.
+    This is an example response.
+    ```json
+    [
+      {
+        "customFieldId": 164,
+        "customFieldParentId": null,
+        "customFieldName": "Checkbox Example",
+        "customFieldGroupId": 64,
+        "customFieldGroupName": "Example Custom Field Group",
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: And here again, an example name that ties together all the fields would be helpful.  The example field names I've chosen all (more or less) line up to an organization whose members are primarily students, so a group name like: "Example Custom Field Group for Student Organizing" would be helpful.

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+        "customFieldTypeId": "M",
+        "isEditable": true,
+        "isRange": true,
+        "maxTextboxCharacters": 0,
+        "availableValues": []
+      },
+      {
+        "customFieldId": 166,
+        "customFieldParentId": null,
+        "customFieldName": "Date Example",
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Possible example here: "Graduation date"

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+        "customFieldId": 164,
+        "customFieldParentId": null,
+        "customFieldName": "Checkbox Example",
+        "customFieldGroupId": 64,
+        "customFieldGroupName": "Example Custom Field Group",
+        "customFieldTypeId": "B",
+        "isEditable": true,
+        "isRange": false,
+        "maxTextboxCharacters": 0,
+        "availableValues": []
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Shouldn't these be null for any field whose type is not "Selection"?

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+        "customFieldTypeId": "D",
+        "isEditable": true,
+        "isRange": true,
+        "maxTextboxCharacters": 0,
+        "availableValues": []
+      },
+      {
+        "customFieldId": 167,
+        "customFieldParentId": null,
+        "customFieldName": "Dropdown Example",
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Possible example here: "Chosen field of study", with available values like "Biology", "Chemistry", "History", etc., below.

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+        "customFieldParentId": null,
+        "customFieldName": "Date Example",
+        "customFieldGroupId": 64,
+        "customFieldGroupName": "Example Custom Field Group",
+        "customFieldTypeId": "D",
+        "isEditable": true,
+        "isRange": true,
+        "maxTextboxCharacters": 0,
+        "availableValues": []
+      },
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: I think it's worth illustrating two Selection examples, to show what a parent-child relationship looks like.  In this case I would suggest constructing an example where one selection is "Department of study", and the other selection is "Field of study", with values that look like this:

* Physical sciences (Department)
** Biology (Field)
** Chemistry (Field)
** Physics (Field)
* Liberal Arts (Department)
** History (Field)
** English (Field)

... or what have you.  The specifics aren't too important, but I think illustrating a good hierarchical selection example is very helpful.

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+        {
+          "id": 5,
+          "name": "Yes",
+          "parentValueId": null
+        }]
+      },
+      {
+        "customFieldId": 168,
+        "customFieldParentId": null,
+        "customFieldName": "Number Example",
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Sample field name: "Number of semesters of study remaining".  Also would be a good place to indicate isRange: true.

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+      },
+      {
+        "customFieldId": 166,
+        "customFieldParentId": null,
+        "customFieldName": "Date Example",
+        "customFieldGroupId": 64,
+        "customFieldGroupName": "Example Custom Field Group",
+        "customFieldTypeId": "D",
+        "isEditable": true,
+        "isRange": true,
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Can ranges be applied to dates?

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+      },
+      {
+        "customFieldId": 165,
+        "customFieldParentId": null,
+        "customFieldName": "Currency Example",
+        "customFieldGroupId": 64,
+        "customFieldGroupName": "Example Custom Field Group",
+        "customFieldTypeId": "M",
+        "isEditable": true,
+        "isRange": true,
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Can ranges be applied to currency?

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 29]
```diff
@@ -0,0 +1,122 @@
...
+        "customFieldTypeId": "N",
+        "isEditable": true,
+        "isRange": true,
+        "maxTextboxCharacters": 75,
+        "availableValues": []
+      },
+      {
+        "customFieldId": 169,
+        "customFieldParentId": null,
+        "customFieldName": "Text Example",
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Example field name: "Name of high school"

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 30]
```diff
@@ -0,0 +1,122 @@
...
+
+block request
+  :markdown
+    This endpoint has no parameters.
+
+block response
+  :markdown
+    This endpoint responds with all Custom Fields that are available in the current context.
+    This is an example response.
+    ```json
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Same comments about id's apply below - please use bogus field ids.  I'd just start with the first one being 500, then 501, 502 ...  Just take care to align the parent and child field ID's for hierarchical selection, and take care to use the same (bogus) group id throughout.

File: docs/van-api/custom-fields/methods/get-customFields.jade [hunk 31]
```diff
@@ -0,0 +1,122 @@
...
+      {
+        "customFieldId": 168,
+        "customFieldParentId": null,
+        "customFieldName": "Number Example",
+        "customFieldGroupId": 64,
+        "customFieldGroupName": "Example Custom Field Group",
+        "customFieldTypeId": "N",
+        "isEditable": true,
+        "isRange": true,
+        "maxTextboxCharacters": 75,
```
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: I thought this property only applied to numeric fields?

File: docs/van-api/people/common-models.jade [hunk 32]
```diff
@@ -79,10 +79,16 @@ block content
           "type": "bsdid",
           "externalId": "746313"
         }
+      ],
+      "customFieldValues": [
+        {
+          "customField": { },
```
Comment (shaisachs) [inline docs/van-api/people/common-models.jade]: This seems wrong to me, shouldn't it be customFieldId?  Furthermore - no point in reporting out a null object.

File: docs/van-api/people/common-models.jade (hunk 8 above)
Comment (shaisachs) [inline docs/van-api/people/common-models.jade]: Yeah definitely

File: docs/van-api/people/methods/md/get-person-response.md [hunk 33]
```diff
@@ -60,7 +60,17 @@ If a result is found, it will look like:
       "isPreferred": null
     }
   ],
-  "identifiers": null
+  "identifiers": null,
+  "customFieldValues": [
+    {
+      "customField": { },
```
Comment (shaisachs) [inline docs/van-api/people/methods/md/get-person-response.md]: Same comment as above re custom field IDs.

File: docs/van-api/custom-fields/methods/get-customFields.jade (hunk 27 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: yes

File: docs/van-api/custom-fields/methods/get-customFields.jade (hunk 28 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: yes

File: docs/van-api/custom-fields/methods/get-customFields.jade (hunk 31 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: i believe it only applies to text fields.

File: docs/van-api/custom-fields/overview.jade (hunk 6 above)
Comment (shaisachs) [inline docs/van-api/custom-fields/overview.jade]: I gave this a bit of a rework and ran it by Torvic: "A Custom Field represents structured data about a person or other object in VAN, which cannot be collected as canvass data or other data-gathering mechanisms.  Custom fields are usually collected for the sole use of a single committee, and generally not shared across committees.  A Contacts Custom Field (which is the only type of custom field available via the API) is a custom field applied to people."

File: docs/van-api/custom-fields/overview.jade [hunk 34]
```diff
@@ -0,0 +1,15 @@
...
+  - var subtopicTitle = "Overview"
+
+block content
+  :markdown
+   Contact Custom Fields allow you to store data that is important to you about a person
+   when standard fields and other data storage options like Activist Codes are not the right tools to store that information.
+   Contact Custom Fields belong to Groups to facilitate organization. For example, you may have a Contact Custom Field Group
+   called &quot;Language&quot; that contains fields such as &quot;Primary Language&quot; and &quot;Speaks English?&quot; to track a person&rsquo;s language preferences.
+   Contact Custom Fields can be any of the following types: Checkbox (boolean), Currency, Date, Dropdown, Number, or Text.
+   Dropdown fields have a set list of acceptable values.
```
Comment (shaisachs) [inline docs/van-api/custom-fields/overview.jade]: I think let's use an example that resonates with the sample response in the GET /customFields doc.  Also a little cleanup:

```
Custom Fields are collected into Groups for better organization and administration.  For example, a committee which works heavily with college students might have a Custom Fields Group called "Student Custom Field Group", with fields such as "Selected field of study", "Name of high school", etc.
```

Then separate out that but about types into a separate paragraph.

```
Every Custom Field has a data type, which indicates what kind of data will be accepted for that custom field.  The data types, and restrictions on the data available, are as follows:

- Boolean: accepts only "true" or false"
- Selection: accepts only one of the available values belonging to the custom field
- Text: accepts any text, limited by the maximum number of characters stipulated in the custom field
- Date: accepts only dates in the format YYYY-MM-DD
- Number: accepts only integers; may be further restricted by range
- Money: accepts currency amounts
```

Comment (shaisachs) [inline docs/van-api/custom-fields/overview.jade]: Finally, we need to answer these questions:
- What format is acceptable for the Date field type?
- Can Number accept negative ints as well as 0/positive?
- What format is acceptable for money - is it just [0-9]+(?\.[0-9]{2})

File: docs/van-api/custom-fields/overview.jade (hunk 7 above)
Comment (shaisachs) [inline docs/van-api/custom-fields/overview.jade]: I think you can just zap this sentence altogether

File: docs/van-api/custom-fields/overview.jade [hunk 35]
```diff
@@ -0,0 +1,15 @@
...
+
```
Comment (shaisachs) [inline docs/van-api/custom-fields/overview.jade]: Also let's not forget our ancestors:

```
Custom Fields of type Selection may have parents to indicate hierarchies within the available values.  The parent of a Custom Field must also be a Selection field, and is usually used to indicate a higher level of groupings for the available values.  If a Custom Field has a parent, then each of its available values also has a parent, and each parent is an available value in the parent custom field. For example, if one custom field indicates "Field of study" for college students, then another might indicate "Department", and their available values might be organized as follows:

- Physical sciences (available value for Department)
-- Biology (available value for Field of study; parent available value is Physical sciences)
-- Chemistry (available value for Field of study; parent available value is Physical sciences)
- Liberal arts (available value for Department)
-- History (available value for Field of study; parent available value is Liberal arts)
-- English (available value for Field of study; parent available value is Liberal arts)
```

File: docs/van-api/custom-fields/methods/get-customFields.jade (hunk 31 above)
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Ok - so we should null it out for this example.

File: docs/van-api/custom-fields/common-models.md (hunk 10 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: Yes. Only integers are valid.

File: docs/van-api/custom-fields/common-models.md (hunk 17 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: Actually, IsRange is a boolean db value we set to true for Number, Dates, and Currency. I don't know why, because the IsRange property is not used anywhere in Van. So I'm going to remove this property from the Dto.

File: docs/van-api/custom-fields/methods/get-customFields.jade (hunk 23 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: yes. i'll fix that.

File: docs/van-api/custom-fields/methods/get-customFields.jade (hunk 31 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: ok. I'll make that field nullable.

File: docs/van-api/custom-fields/common-models.md (hunk 17 above)
Comment (shaisachs) [inline docs/van-api/custom-fields/common-models.md]: Awesome, makes sense!

File: docs/van-api/custom-fields/methods/get-customFields.jade (hunk 23 above)
Comment (shaisachs) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: Woo, thanks

File: docs/van-api/people/common-models.jade (hunk 8 above)
Comment (shaisachs) [inline docs/van-api/people/common-models.jade]: Question courtesy of Torvic: if you set assignedValue to null, will that erase the value?

**Comment by shaisachs** [comment]
Basically my view on Custom Fields on the /people model is that they should behave very similarly to the way Locations on the /events model work:
- the /people common object section should, in the description area, describe the custom fields section as "A simple Custom Field which contains the assigned value" or some such
- the example of a /people common object should indicate a fully fleshed-out custom field alongside an assigned value

That's pretty close to what we already have, I just want to make sure we have it all nice and consistent.

File: docs/van-api/people/common-models.jade (hunk 8 above)
Comment (ericmontag) [inline docs/van-api/people/common-models.jade]: Nina asked the same question. If the assignedValue is an empty string, the db custom field value will be set to NULL.
I need to test and make sure the code does the same if assignedValue is null.

Comment (shaisachs) [inline docs/van-api/people/common-models.jade]: Great, thanks!

File: docs/van-api/people/common-models.jade (hunk 32 above)
Comment (ericmontag) [inline docs/van-api/people/common-models.jade]: Is it ok to require custom field group id as well?
If not, we will need to get the group id from the database, which is what we do for Online Forms.

Comment (shaisachs) [inline docs/van-api/people/common-models.jade]: I would suggest not requiring the custom field group id if at all possible, since it increases the chance of something going wrong (e.g. a request that sends a custom field group id which does not correspond to the group for the the custom field id).  OTOH - I guess it's not the end of the world.  There are some comparable places where we require two IDs when one would suffice.

Also, see my most recent comment - I think it's ok to provide the full custom field object in the GET /people/1234 response.  I would just provide a full custom object in the example, here.  Otherwise it looks like we're always kicking back empty objects, which is misleading.

Comment (ericmontag) [inline docs/van-api/people/common-models.jade]: Then can we require Online Forms to also send the custom field group id, so we don't need to retrieve it from the database?

Comment (shaisachs) [inline docs/van-api/people/common-models.jade]: Can you email that over to Sarah Cartwright and Kevin Thayer, and cc me?  I don't know the answer, really.

File: docs/van-api/people/common-models.jade (hunk 8 above)
Comment (ericmontag) [inline docs/van-api/people/common-models.jade]: If the assignedValue is null, the db custom field value will be set to NULL.

File: docs/van-api/custom-fields/common-models.md (hunk 10 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: Only positive integers are valid.

File: docs/van-api/custom-fields/common-models.md (hunk 1 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: ok

File: docs/van-api/custom-fields/common-models.md (hunk 2 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: ok

File: docs/van-api/custom-fields/common-models.md (hunk 3 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: ok

File: docs/van-api/custom-fields/common-models.md (hunk 4 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: ok

File: docs/van-api/custom-fields/methods/get-customFields.jade (hunk 5 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/methods/get-customFields.jade]: fixed

File: docs/van-api/custom-fields/overview.jade (hunk 7 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/overview.jade]: ok

File: docs/van-api/people/common-models.jade (hunk 8 above)
Comment (ericmontag) [inline docs/van-api/people/common-models.jade]: Changed wording to 'The value must be consistent'

File: docs/van-api/people/methods/md/get-person-response.md (hunk 9 above)
Comment (ericmontag) [inline docs/van-api/people/methods/md/get-person-response.md]: Actually we were returning the raw database value, "0" or "1".

Fixed this to return a string, "true" or "false".

File: docs/van-api/custom-fields/common-models.md (hunk 16 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: ok

File: docs/van-api/custom-fields/common-models.md (hunk 18 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: ok

File: docs/van-api/custom-fields/common-models.md (hunk 14 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: ok

File: docs/van-api/custom-fields/overview.jade (hunk 6 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/overview.jade]: changed the text of the Overview

File: docs/van-api/custom-fields/common-models.md (hunk 15 above)
Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: ok

**Comment by ericmontag** [comment]
@shaisachs @BaconSoap  @NGPVAN/falcon 
Please review. These are changes from the review comments.

File: docs/van-api/people/common-models.jade [hunk 36]
```diff
@@ -79,10 +79,17 @@ block content
           "type": "bsdid",
           "externalId": "746313"
         }
+      ],
+      "customFieldValues": [
+        {
+          "customFieldId": 500,
+          "customFieldGroupId: 100,
```
Comment (shaisachs) [inline docs/van-api/people/common-models.jade]: Missing a closing quote (") here

**Comment by shaisachs** [comment]
Lookin good!  Other than that one typo I think we're in good shape!

File: docs/van-api/people/common-models.jade (hunk 36 above)
Comment (ericmontag) [inline docs/van-api/people/common-models.jade]: fixed

File: docs/van-api/custom-fields/overview.jade [hunk 37]
```diff
@@ -0,0 +1,40 @@
...
+    called &quot;Student Custom Field Group&quot;, with fields such as &quot;Selected field of study&quot;, &quot;Name of high school&quot;, etc.
+
+    Every Custom Field has a data type, which indicates what kind of data will be accepted for that custom field.
+    The data types, and restrictions on the data available, are as follows:
+
+    - Boolean: accepts only &quot;true&quot; or &quot;false&quot;
+    - Selection: accepts only one of the available values belonging to the custom field
+    - Text: accepts any text, limited by the maximum number of characters stipulated in the custom field
+    - Date: accepts only dates in the format YYYY-MM-DD
+    - Number: accepts only positive integers and zero
```
Comment (shaisachs) [inline docs/van-api/custom-fields/overview.jade]: I know this is annoying, but, let's change this to "accepts only non-negative integers"

Comment (ericmontag) [inline docs/van-api/custom-fields/overview.jade]: no problem - it's not annoying.

Comment (ericmontag) [inline docs/van-api/custom-fields/overview.jade]: changed.

File: docs/van-api/custom-fields/common-models.md [hunk 38]
```diff
@@ -0,0 +1,63 @@
+
+* CustomFieldTypes:
+    * **B** = Boolean (Checkbox)
+    * **S** = Selection (Dropdown)
+    * **T** = Text
+    * **D** = Date
+    * **N** = Number (Positive integer)
```
Comment (girlprogrammer) [inline docs/van-api/custom-fields/common-models.md]: It's my understanding that 0 is allowed, making this a nonnegative integer, not a positive integer.

Comment (ericmontag) [inline docs/van-api/custom-fields/common-models.md]: Changed this to non-negative integer

**Comment by shaisachs** [comment]
@ericmontag @jackdolabany @JacksonianDemocracy @itsdrewmiller 
Planning to push this live.  You can preview the changes in question at http://ngpvan-dev-portal-feature.herokuapp.com/, so please give it a once-over (and feel free to cc in others who may be interested).  If all looks well we'll ship it!
